File: backend/src/services/table.service.ts
import crypto from 'crypto';
import { AppError } from '../utils/AppError.js';

// ─── Config ───────────────────────────────────────────────────────────────────

const MAX_TABLE_NUMBER_LENGTH = 10;

function getSecret(): string {
  const secret = process.env.QR_SECRET;
  if (!secret) throw new AppError(500, 'QR_SECRET chưa được cấu hình');
  return secret;
}

// ─── Service ──────────────────────────────────────────────────────────────────

export const tableService = {
  /**
   * Validates that a table number is non-empty and within length limits.
   * Throws AppError(400) on failure.
   */
  validateTableNumber(tableNumber: string): void {
    if (!tableNumber || tableNumber.trim().length === 0) {
      throw new AppError(400, 'Số bàn không được để trống');
    }
    if (tableNumber.length > MAX_TABLE_NUMBER_LENGTH) {
      throw new AppError(
        400,
        `Số bàn không được quá ${MAX_TABLE_NUMBER_LENGTH} ký tự`,
      );
    }
  },

  /**
   * Generates a 16-char lowercase hex HMAC token for a table number.
   * Token is deterministic: same table + same QR_SECRET → same token.
   * To invalidate all QR codes, change QR_SECRET in .env.
   */
  generateToken(tableNumber: string): string {
    return crypto
      .createHmac('sha256', getSecret())
      .update(tableNumber)
      .digest('hex')
      .slice(0, 16);
  },

  /**
   * Timing-safe comparison of the provided token against the expected HMAC.
   * Returns false (instead of throwing) on any mismatch or invalid input.
   */
  verifyToken(tableNumber: string, token: string): boolean {
    if (!token || token.length < 16) return false;
    try {
      const expected = this.generateToken(tableNumber);
      const provided = token.slice(0, 16);
      // Both strings are 16 ASCII chars → 16-byte UTF-8 buffers (same length)
      return crypto.timingSafeEqual(
        Buffer.from(expected, 'utf8'),
        Buffer.from(provided, 'utf8'),
      );
    } catch {
      return false;
    }
  },

  /**
   * Builds the full QR URL for a table.
   * Format: {FRONTEND_URL}/scan?table={tableNumber}&sid={token}
   */
  generateQrUrl(tableNumber: string): string {
    const token = this.generateToken(tableNumber);
    const base = (process.env.FRONTEND_URL || 'http://localhost:5173').replace(
      /\/$/,
      '',
    );
    return `${base}/scan?table=${encodeURIComponent(tableNumber)}&sid=${token}`;
  },
};

File: backend/app/services/table_service.py
import hmac
import hashlib
from app.utils.app_error import AppError
from app.config import settings

MAX_TABLE_NUMBER_LENGTH = 10

def _get_secret() -> str:
    if not settings.QR_SECRET:
        raise AppError(500, "QR_SECRET chưa được cấu hình")
    return settings.QR_SECRET

def validate_table_number(table_number: str) -> None:
    if not table_number or not table_number.strip():
        raise AppError(400, "Số bàn không được để trống")
    if len(table_number) > MAX_TABLE_NUMBER_LENGTH:
        raise AppError(400, f"Số bàn không được quá {MAX_TABLE_NUMBER_LENGTH} ký tự")

def generate_token(table_number: str) -> str:
    secret = _get_secret()
    return hmac.new(secret.encode(), table_number.encode(), hashlib.sha256).hexdigest()[:16]

def verify_token(table_number: str, token: str) -> bool:
    if not token or len(token) < 16:
        return False
    try:
        expected = generate_token(table_number)
        provided = token[:16]
        return hmac.compare_digest(expected.encode(), provided.encode())
    except Exception:
        return False

def generate_qr_url(table_number: str) -> str:
    token = generate_token(table_number)
    base = settings.FRONTEND_URL.rstrip("/")
    from urllib.parse import quote
    return f"{base}/scan?table={quote(table_number)}&sid={token}"
